services/headless-lms/chatbot/src/chatbot_tools/course_progress.rs

```diff
@@ -0,0 +1,172 @@
+    fn output(&self) -> String {
+        let mut progress = self.state.progress.to_owned();
+        let course_name = &self.state.course_name;
+        let mut res = format!("The user is completing a course called {course_name}. ");
+
+        if progress.len() == 1 {
+            let module = &progress[0];
+            res.push_str("Their progress on this course is the following: ");
+
+            res = push_exercises_scores_progress(
+                res,
+                module.attempted_exercises,
+                module.total_exercises,
+                module.attempted_exercises_required,
+                module.score_given,
+                module.score_maximum,
+                module.score_required,
+            );
+        } else {
+            progress.sort_by_key(|m| m.course_module_order_number);
+            let first_mod = progress.first();
+
+            res = if let Some(module) = first_mod {
+                let m_name = &module.course_module_name;
+                res.push_str(&format!(
+                "The user's progress on the base course module called {m_name} is the following: "
+            ));
+                push_exercises_scores_progress(
+                    res,
+                    module.attempted_exercises,
+                    module.total_exercises,
+                    module.attempted_exercises_required,
+                    module.score_given,
+                    module.score_maximum,
+                    module.score_required,
+                )
+            } else {
+                res.push_str("There is no progress information for this user on this course. ");
+                res
+            };
+            for module in progress.iter().skip(1) {
+                let m_name = &module.course_module_name;
+                res.push_str(&format!(
+                    "The user's progress on the course module called {m_name} is the following: "
+                ));
+                res = push_exercises_scores_progress(
+                    res,
+                    module.attempted_exercises,
+                    module.total_exercises,
+                    module.attempted_exercises_required,
+                    module.score_given,
+                    module.score_maximum,
+                    module.score_required,
+                );
+            }
+        }
+        res
+    }
+
+    fn get_arguments(&self) -> &Self::Arguments {
+        &self.arguments
+    }
+}
+
+#[derive(serde::Serialize)]
+pub struct CourseProgressArguments {}
+
+pub struct CourseProgressState {
+    course_name: String,
+    progress: Vec<UserCourseProgress>,
+}
+
+pub async fn call_chatbot_tool(
+    conn: &mut PgConnection,
+    fn_name: &str,
+    fn_args: &str,
+    user_context: &ChatbotUserContext,
+) -> anyhow::Result<impl ChatbotTool> {
+    let output = match fn_name {
+        // if the function took arguments, this is how:
+        /* "foo" => {
```

Comment:
dead code (you can, however, document these kind of things in the trait documentation)